runtime/signals.c

```diff
@@ -277,40 +278,33 @@ void caml_update_young_limit (void)
 void caml_request_major_slice (void)
 {
   Caml_state->requested_major_slice = 1;
-  caml_set_something_to_do();
+  caml_set_action_pending();
 }
 
 void caml_request_minor_gc (void)
 {
   Caml_state->requested_minor_gc = 1;
-  caml_set_something_to_do();
+  caml_set_action_pending();
 }
 
-value caml_check_gc_without_async_callbacks(value extra_root)
+void caml_process_pending_callbacks(void)
 {
-  CAMLparam1 (extra_root);
-  if (Caml_state->requested_major_slice || Caml_state->requested_minor_gc){
-    CAML_INSTR_INT ("force_minor/check_gc_without_async_callbacks@", 1);
-    caml_gc_dispatch();
-  }
-  CAMLreturn (extra_root);
-}
-
-CAMLexport value caml_check_urgent_gc (value extra_root)
-{
-  CAMLparam1 (extra_root);
   caml_something_to_do = 0;
   caml_update_young_limit();
-  if (Caml_state->requested_major_slice || Caml_state->requested_minor_gc){
-    CAML_INSTR_INT ("force_minor/check_urgent_gc@", 1);
-    caml_gc_dispatch();
-  }
-  caml_memprof_handle_postponed();
+  caml_memprof_handle_postponed(Val_unit);
   caml_final_do_calls();
   caml_process_pending_signals();
-  CAMLreturn (extra_root);
 }
 
+CAMLno_tsan /* The access to [caml_something_to_do] is not synchronized. */
+CAMLexport void caml_process_pending_actions(void)
```

Comment:
Perhaps we could make this an `inline` function in the `.h` file for better performance? Of course, that would mean that we need to make public `caml_something_to_do` and `caml_process_prinding_callbacks`. But we can document that they are effectively private functions.

Reply:
I need to learn about inlining in C first. Any obstacle to postponing?

Reply:
There is nothing too fancy about this: you just put the source of the function it in the `.h` file, with the `inline` keyword. The function is going to be recompiled every time the `.h` file is included, so you have to make sure it is well-defined (i.e., it does not use undefined symbols) everywhere the `.h` is going to be included (including when `CAML_INTERNALS` is not defined).

Reply:
Ah, and, in order to follow the style in other `.h` files, you should also make the function `static`.

Reply:
I'll have a look, later.

Reply:
What I could learn from various sources is that one should put the "inline" function definition in the header, and an "extern inline" function declaration in the .c file.

In any case, this looks like a premature optimisation that requires us to cheat about what is internal or not,  so I prefer not to do it.